crates/engine/tree/docs/root.md

```diff
@@ -0,0 +1,244 @@
+# State Root Calculation for Engine Payloads
+
+The heart of Reth is the Engine, which is responsible for driving the chain forward.
+Each time it receives a new payload ([engine_newPayloadV4](https://github.com/ethereum/execution-apis/blob/main/src/engine/prague.md#engine_newpayloadv4)
+at the time of writing this document), it:
+1. Does a bunch of validations
+2. Executes the block contained in the payload
+3. Calculates the [MPT](https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/)
+root of the new state
+4. Compares the root with the one received in the block header,
+5. Considers the block committed.
+
+This document describes the lifecycle of a payload with the focus on state root calculation,
+from the moment the payload is received, to the moment we have a new state root.
+
+We will look at the following components:
+- [Engine](#engine)
+- [State Root Task](#state-root-task)
+- [MultiProof Manager](#multiproof-manager)
+- [Sparse Trie Task](#sparse-trie-task)
+
+## Engine
+
+![Engine](./mermaid/engine.mmd.svg)
+
+It all starts with the `engine_newPayload` request coming from the [Consensus Client](https://ethereum.org/en/developers/docs/nodes-and-clients/#consensus-clients).
+
+We extract the block from the payload, and eventually pass it to the `EngineApiTreeHandler::insert_block_inner`
+method, https://github.com/paradigmxyz/reth/blob/2ba54bf1c1f38c7173838f37027315a09287c20a/crates/engine/tree/src/tree/mod.rs#L2359-L2362
+
+First, we spawn the [State Root Task](#state-root-task) thread, which will receive the updates from
+execution and calculate the state root. https://github.com/paradigmxyz/reth/blob/2ba54bf1c1f38c7173838f37027315a09287c20a/crates/engine/tree/src/tree/mod.rs#L2449-L2458
+
+Then, we do two things with the block:
+1. Start prewarming each transaction in a separate thread ("Prewarming thread" on the above diagram).
+https://github.com/paradigmxyz/reth/blob/2ba54bf1c1f38c7173838f37027315a09287c20a/crates/engine/tree/src/tree/mod.rs#L2490-L2507
+    - Each transaction is optimistically executed in parallel with each other on top of the previous block,
+    but the results are not committed to the database.
+    - All accounts and storage slots that were accessed are cached in memory, so that the actual execution
+    can use them instead of going to the disk.
+    - All modified accounts and storage slots are sent as `StateRootMessage::PrefetchProofs`
+    to the [State Root Task](#state-root-task).
+    - Some transactions will fail, because they require the previous transactions to be executed first.
```

Comment:
maybe add: but it doesn't matter that they fail, bc now we have them in memory and can re-execute them quickly